torch/nn/functional.py

```diff
@@ -3953,3 +3953,186 @@ def multi_head_attention_forward(query,                           # type: Tensor
+        num_heads (int): number of parallel heads used.
+        in_proj_weight (Tensor): weight used for projection
+        in_proj_bias (Tensor, optional): bias used for projection.
+
+    Shape:
+        - seq: :math:`(S, N, E)`
+        - in_proj_weight: :math:`(P, E)`
+        - in_proj_bias: :math:`(P)`
+        - Output: :math:`(N * H, S, P / H)`
+        where S is the sequence length, H is the number of attention heads, N is the
+        batch size, P is the projection dimension, and E is the embedding
+        dimension.
+
+    """
+    if not torch.jit.is_scripting():
+        tens_ops = (seq, in_proj_weight)
+        if any([type(t) is not Tensor for t in tens_ops]) and has_torch_function(tens_ops):
+            return handle_torch_function(
+                multi_head_attention_in_projection, tens_ops,
+                seq, num_heads, in_proj_weight, in_proj_bias=in_proj_bias)
+    seq_len, bsz, _ = seq.size()
+    proj_dim = in_proj_weight.size(0)
+    assert proj_dim % num_heads == 0, "projection dimension must be divisible by num_heads"
+    head_dim = proj_dim // num_heads
+
+    q = linear(seq, in_proj_weight, in_proj_bias)
+    # Shape of q: (S, N, P)
+    q = q.reshape(seq_len, bsz * num_heads, head_dim).transpose(0, 1)
+    return q
+
+
+def scaled_dot_product_attention(q,                         # type: Tensor
+                                 k,                         # type: Tensor
+                                 v,                         # type: Tensor
+                                 num_heads,                 # type: int
+                                 add_zero_attn,             # type: bool
+                                 dropout_p,                 # type: float
+                                 training=True,             # type: bool
+                                 key_padding_mask=None,     # type: Optional[Tensor]
+                                 attn_mask=None,            # type: Optional[Tensor]
+                                 ):
+    # type: (...) -> Tuple[Tensor, Tensor]
+    r"""Uses a scaled dot product with the projected key-value pair to update
+    the projected query.
+
+    Args:
+        q (Tensor): Projected query
+        k (Tensor): Projected key
+        v (Tensor): Projected value
+        num_heads (int): Number of parallel attention heads.
+        add_zero_attn (bool): Add a new batch of zeros to the projected key and
+            value sequences at dimension 1.
+        dropout_p (float): Probability of an element will be zeroed.
+        training (bool): Apply dropout if ``training=True``
+        key_padding_mask (Tensor, optional): Specified padding elements in the
+            key will be ignored by the attention. This is a binary mask. When
+            the value is True, the corresponding value on the attention layer
+            will be set to :math:`-\inf`.
+        attn_mask (Tensor, optional): 2D or 3D mask that prevents attention to
```

Comment:
I have a PR to add bool tensor support for `attn_mask` and eventually, we will switch to bool tensor for both `attn_mask` and `key_padding_mask`. For those building blocks, I think we should start with bool tensor for the masks.
https://github.com/pytorch/pytorch/pull/33763

Reply:
I can work on adding that once the PR lands.

Reply:
No worry for that PR. I will make sure it land. 
For this PR, I think it makes sense to use bool tensor for `attn_mask` and `key_padding_mask`.

Reply:
Okay. I have also included support for a boolean `attn_mask`. `key_padding_mask` was already a boolean tensor.

Reply:
I think it makes sense to support only bool tensor for `attn_mask` and `key_mask`. We have both float and bool support for `attn_mask` in MHA because the `attn_mask` was initially implemented as float tensor. We don't need to do that again for those new building blocks as long as we make it clear in the doc.

Reply:
Okay. I have an update that changes the documentation and removes the type-dependent behavior.